src/headless/converse-chat.js

```diff
@@ -1100,7 +1100,17 @@ converse.plugins.add('converse-chat', {
                     return;
                 }
                 if (utils.isNewMessage(message) && this.isHidden()) {
-                    this.save({'num_unread': this.get('num_unread') + 1});
+                    let first_unread = this.get('first_unread');
+
+                    if (this.get('num_unread') == 0) {
+                      if (first_unread) {
+                          const msg = this.messages.where({'msgid': first_unread});
```

Comment:
`msgid` isn't guaranteed to be unique. This is the `id` attribute that the client puts on the stanza, and older clients in particular don't maintain globally unique message ids.

You should instead try to use the XEP-359 stanza id if it's available, otherwise try the `origin_id` and lastly fall back to the `msgid`.

The stanza ID needs to be checked together with the `by` value, something similar to what's being done here:
https://github.com/conversejs/converse.js/blob/29638309168551d609a797e5e17e8e214d695955/src/headless/converse-mam.js#L41

If the stanza id isn't available, then you can probably fall back to to the `msgid`.

---

You can use `messages.findWhere` instead of `where` to stop once the first match has been found and you can use the new optional chaining syntax to shorten everything into one line.

`this.messages.findWhere(query)?.save("first_unread", false);`

Also, is there a particular reason why you're using `set` and not `save`? With `set`, the values aren't persisted to localStorage.

Reply:
> `msgid` isn't guaranteed to be unique.

We don't need this to be unique because it is not persistent. The requirement is for real-time messages only. When messages are re-fetched from MAM or persistent storage, their first_unread message indicator is reset to false.

Reply:
I think this information should be persisted in localStorage, otherwise when you reload the tab you'll lose the unread marker, even if you haven't looked at the messages yet.

To simplify things (if we don't need to store an attribute from the actual stanza), you can just save the `id` attribute instead of the `msgd`. That's also unique and then you can do a direct lookup instead of using `findWhere` you can just use `this.messages.get(id)`.

Reply:
Thanks. I will  use id instead of msgid